Exercises.AsyncAwait/UrlSaver.cs

```diff
@@ -0,0 +1,163 @@
+﻿using Exercises.AsyncAwait.Dependences;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Exercises.AsyncAwait
+{
+    public sealed class UrlSaver
+    {
+        private readonly IStorage _storage;
+        private readonly ILogger _logger;
+
+        private readonly TaskCompletionSource<int> _taskCompletionSource;
+        private readonly ConcurrentQueue<string> _queue;
+
+        private IDictionary<string, string> _urls;
+        private decimal _quantumOfStatusLine;
+        private int _workingTasks;
+        private object _locker;
+        private int _counter;
+
+        public event EventHandler<StatusLineEventArgs> UpdateStatusLine;
+
+
+        public UrlSaver(IStorage storage, ILogger logger)
+        {
+            _storage = storage;
+            _logger = logger;
+            _urls = new Dictionary<string, string>();
+            _queue = new ConcurrentQueue<string>();
+            _workingTasks = 0;
+            _taskCompletionSource = new TaskCompletionSource<int>();
+            _counter = 0;
+            _locker = new Object();
+        }
+
+
+        public async Task<bool> TryGetGetUrlsFromStorageAsync()
+        {
+            _urls = await _storage.GetUrlsFromStorageAsync();
+            return _urls.Count > 0;
+        }
+
+        public async Task<int> GetDataFromUrlAsync(int downloadsQuantity)
+        {
+            var t = _taskCompletionSource.Task;
+
+            foreach (var url in _urls)
+            {
+                _queue.Enqueue(url.Key);
+            }
+            _quantumOfStatusLine = 100m / _queue.Count;
+            _logger.Write($"In queue {_queue.Count} downloads");
+
+            var tasksInTimeQuantity = Math.Min(_urls.Count, downloadsQuantity);
+
+            for (int i = 0; i < tasksInTimeQuantity; i++)
+            {
+                TaskRunAsync();
+            }
+
+            return await t;
+        }
+
+        public async Task SaveContentToStorageAsync()
+        {
+            foreach (var url in _urls)
+            {
+                await _storage.SaveContentToStorageAsync(url.Key, url.Value);
+            }
+        }
+
+
+        private async void TaskRunAsync()
+        {
+            _workingTasks++;
```

Comment:
If write here next lines, will your program notice this exception?
```suggestion
            throw new Exception("PWND!!!");
```